mart/utils/modality_dispatch.py

```diff
@@ -0,0 +1,68 @@
+#
+# Copyright (C) 2022 Intel Corporation
+#
+# SPDX-License-Identifier: BSD-3-Clause
+#
+
+from __future__ import annotations
+
+from itertools import cycle
+from typing import Any, Callable, Iterable
+
+import torch
+
+__all__ = ["modality_dispatch"]
+
+DEFAULT_MODALITY = "default"
+
+
+def modality_dispatch(
+    modality_func: Callable | dict[str, Callable],
+    data: torch.Tensor | dict[str, torch.Tensor] | Iterable[Any],
+    *,
+    input: torch.Tensor | dict[str, torch.Tensor] | Iterable[Any],
+    target: torch.Tensor | Iterable[Any] | None,
+    modality: str = DEFAULT_MODALITY,
+):
+    """Recursively dispatch data and input/target to functions of the same modality.
+
+    The function returns an object that is homomorphic to input and data.
+    """
+
+    assert type(data) == type(input)
+    if target is None:
+        # Make target zips well with input.
+        target = cycle([None])
+
+    if isinstance(input, torch.Tensor):
+        if isinstance(modality_func, dict):
+            # A dictionary of Callable indexed by modality.
+            return modality_func[modality](data, input=input, target=target)
+        else:
+            # A Callable with modality=? as a keyword argument.
+            return modality_func(data, input=input, target=target, modality=modality)
+    elif isinstance(input, dict):
+        # The dict input has modalities specified in keys, passing them recursively.
+        output = {}
+        for modality in input.keys():
+            output[modality] = modality_dispatch(
+                modality_func,
+                data[modality],
+                input=input[modality],
+                target=target,
+                modality=modality,
+            )
+        return output
+    elif isinstance(input, (list, tuple)):
```

Comment:
Use Iterable

Reply:
I can't, because `dict` is also `Iterable`.

Reply:
But you already check `dict` above? I guess this is fine assuming the type is `Iterable` above.